stateproof/builder.go

```diff
@@ -20,22 +20,142 @@ import (
 	"database/sql"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"sort"
 
 	"github.com/algorand/go-algorand/config"
 	"github.com/algorand/go-algorand/crypto/stateproof"
 	"github.com/algorand/go-algorand/data/basics"
-	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/stateproofmsg"
 	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
 	"github.com/algorand/go-algorand/logging"
 	"github.com/algorand/go-algorand/network"
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-algorand/stateproof/verify"
 )
 
-// makeBuilderForRound not threadsafe, should be called in a lock environment
-func (spw *Worker) makeBuilderForRound(rnd basics.Round) (builder, error) {
+var errVotersNotTracked = errors.New("voters not tracked for the given lookback round")
+
+// OnPrepareVoterCommit is a function called by the voters tracker when it's preparing to commit rnd. It gives the builder
+// the chance to persist the data it needs.
+func (spw *Worker) OnPrepareVoterCommit(rnd basics.Round, votersFetcher ledgercore.VotersForRoundFetcher) error {
+	header, err := spw.ledger.BlockHdr(rnd)
+	if err != nil {
+		return fmt.Errorf("OnPrepareVoterCommit(%d): could not fetch round header: %w", rnd, err)
+	}
+
+	proto := config.Consensus[header.CurrentProtocol]
+	if proto.StateProofInterval == 0 || uint64(rnd)%proto.StateProofInterval != 0 {
+		return nil
+	}
+
+	builderExists, err := spw.builderExists(rnd)
+	if err != nil {
+		spw.log.Warnf("OnPrepareVoterCommit(%d): could not check builder existence, assuming it doesn't exist: %v\n", rnd, err)
+	} else if builderExists {
+		return nil
+	}
+
+	// We create and persist the builder without using it. Signer can later use it (specifically the state proof
+	// message embedded in the builder) during catchup.
+	_, err = spw.createAndPersistBuilder(rnd, votersFetcher)
+	if err != nil {
+		if errors.Is(err, errVotersNotTracked) {
+			spw.log.Warnf("OnPrepareVoterCommit(%d): %v", rnd, err)
+			return nil
+		}
+
+		return fmt.Errorf("OnPrepareVoterCommit(%d): could not create builder: %w", rnd, err)
+	}
+
+	return nil
+}
+
+// loadOrCreateBuilderWithSignatures either loads a builder from the DB or creates a new builder.
+// this function fills the builder with all the available signatures
+func (spw *Worker) loadOrCreateBuilderWithSignatures(rnd basics.Round) (builder, error) {
+	b, err := spw.loadOrCreateBuilder(rnd)
+	if err != nil {
+		return builder{}, err
+	}
+
+	if err := spw.loadSignaturesIntoBuilder(&b); err != nil {
+		return builder{}, err
+	}
+	return b, nil
+}
+
+func (spw *Worker) loadOrCreateBuilder(rnd basics.Round) (builder, error) {
+	buildr, err := spw.loadBuilderFromDB(rnd)
+	if err == nil {
+		return buildr, nil
+	}
+
+	if !errors.Is(err, sql.ErrNoRows) {
+		spw.log.Errorf("loadOrCreateBuilder: error while fetching builder from DB: %v", err)
+	}
+
+	buildr, err = spw.createAndPersistBuilder(rnd, spw.ledger)
```

Comment:
this presumably should not happen at all since we creating builders are we commit, and the sig messages are expected to come after. Maybe log/count such situations?

Reply:
it might happen. Usually, when catching up, a node will use the commit callback to create a builder.
when synced, stateproof signer will create the builder.